attestation/link/link.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2024 The Witness Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package link
+
+import (
+	"encoding/json"
+	"fmt"
+
+	v0 "github.com/in-toto/attestation/go/predicates/link/v0"
+	v1 "github.com/in-toto/attestation/go/v1"
+	"github.com/in-toto/go-witness/attestation"
+	"github.com/in-toto/go-witness/attestation/commandrun"
+	"github.com/in-toto/go-witness/attestation/environment"
+	"github.com/in-toto/go-witness/attestation/material"
+	"github.com/in-toto/go-witness/attestation/product"
+	"github.com/in-toto/go-witness/cryptoutil"
+	"github.com/in-toto/go-witness/registry"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+const (
+	Name    = "link"
+	Type    = "https://in-toto.io/attestation/link/v0.3"
+	RunType = attestation.PostProductRunType
+
+	defaultExport = false
+)
+
+// This is a hacky way to create a compile time error in case the attestor
+// doesn't implement the expected interfaces.
+var (
+	_ attestation.Attestor  = &Link{}
+	_ attestation.Subjecter = &Link{}
+)
+
+func init() {
+	attestation.RegisterAttestation(Name, Type, RunType,
+		func() attestation.Attestor { return New() },
+		registry.BoolConfigOption(
+			"export",
+			"Export the link attestation to its own file",
+			defaultExport,
+			func(a attestation.Attestor, export bool) (attestation.Attestor, error) {
+				linkAttestor, ok := a.(*Link)
+				if !ok {
+					return a, fmt.Errorf("unexpected attestor type: %T is not a link attestor", a)
+				}
+				WithExport(export)(linkAttestor)
+				return linkAttestor, nil
+			},
+		),
+	)
+}
+
+type Option func(*Link)
+
+func WithExport(export bool) Option {
+	return func(l *Link) {
+		l.export = export
+	}
+}
+
+type Link struct {
+	PbLink   v0.Link
+	products map[string]attestation.Product
+	export   bool
+}
+
+func New() *Link {
+	return &Link{}
+}
+
+func (l *Link) Name() string {
+	return Name
+}
+
+func (l *Link) Type() string {
+	return Type
+}
+
+func (l *Link) RunType() attestation.RunType {
+	return RunType
+}
+
+func (l *Link) Export() bool {
+	return l.export
+}
+
+func (l *Link) Attest(ctx *attestation.AttestationContext) error {
+	l.PbLink.Name = "stepNameHere"
+	for _, attestor := range ctx.CompletedAttestors() {
```

Comment:
Am I right in saying that this *must* run last? I know we're running this as a "PostProduct", but could we run the risk of not catching other "PostProduct" attestors? Maybe this functionality would be better served as a flag that generates a link attestation rather than an attestor type?

Reply:
I was thinking that a new lifecycle type of "composite" that runs after all the other attestors. The composite attestor could be given a copy of the data instead of direct references. That wouldn't allow someone to "accidentally" modify the results of any other attestor.

Reply:
Yeah I definitely think something that handled stuff **after** the attestors run would be the way to go here. It also keeps things decoupled as you say, to in the future completely lock attestors out from seeing each others metadata.

Reply:
would it make sense for the link attester to be post product?  I'm concerned about adding more steps to the lifecycle, and making the user define the order would not be ideal

Reply:
> would it make sense for the link attester to be post product
We actually need it to be the last post product. Example would be to add a subject for an OCI image. We have to run after the OCI post product attestor.

Reply:
Seems it might be sensible to (at least for now) ensure that this attestor is handled last. I am also guessing that this sort of behaviour would need to apply to the SLSA attestor that sits in this PR. 

We should probably think more about the seemingly increasing number of attestors that will not gather information themselves but instead rely on evidence generated by others. Do we have an existing issue for this @jkjell @mikhailswift ?

Reply:
created one https://github.com/in-toto/witness/issues/422. Don't think we need to implement this in order to have this stuff merged though.